Add Imagewang dataset

rickwierenga opened the pull request:
See https://github.com/tensorflow/datasets/issues/1325 for more details.

WilliamHYZhang commented on tensorflow_datasets/image/imagewang_test.py:
```diff
@@ -0,0 +1,19 @@
+""" Tests for Imagewang contains Imagenette and Imagewoof combined. """
```
Consider creating 3 different tests (different classes) for the 3 variants of the dataset. Usually 1 file for each split is enough.

WilliamHYZhang replied:
Use `BUILDER_CONFIG_NAMES_TO_TEST` and set it equal to `['<name of builder config>']` when implementing the classes.

rickwierenga replied:
Is this better: https://github.com/tensorflow/datasets/pull/1326/commits/3f0754ecb8fca4a575f93b82e514e0dd945be22a?

The tests fail because there is no test split. The reason for that is because Imagewang does not have a publicly available test set.

WilliamHYZhang commented on tensorflow_datasets/image/imagewang.py:
```diff
@@ -0,0 +1,142 @@
+    if size in _SIZES:
+      size_str = "" if size == "full-size" else "-" + size[:-2]
+      url = os.path.join(_URL_PREFIX, "imagewang%s.tgz" % size_str)
+      print(url)
```
Consider removing this

WilliamHYZhang commented on tensorflow_datasets/image/imagewang.py:
```diff
@@ -0,0 +1,142 @@
+      train_path = os.path.join(path, _SIZE_TO_DIRNAME[size], "train")
+      val_path = os.path.join(path, _SIZE_TO_DIRNAME[size], "val")
+    else:
+      raise ValueError("Size not implemented!")
```
Can this be changed to something more informative, i.e, `raise ValueError("size must be one of %s" % _SIZES)`?

WilliamHYZhang commented on tensorflow_datasets/image/imagewang_test.py:
```diff
@@ -0,0 +1,19 @@
+class ImagewangTest(testing.DatasetBuilderTestCase):
+  DATASET_CLASS = imagewang.Imagewang
+  SPLITS = {
+      "train": 4,  # Number of fake train example
```
Consider removing comment

WilliamHYZhang commented on tensorflow_datasets/image/imagewang_test.py:
```diff
@@ -0,0 +1,19 @@
+  DATASET_CLASS = imagewang.Imagewang
+  SPLITS = {
+      "train": 4,  # Number of fake train example
+      "test": 4,  # Number of fake test example
```
Consider removing comment

rickwierenga commented:
Thanks for the review! I left some stupid code in there, glad you caught it. All issues except for the one I commented on should be resolved.

> and good luck in GCI!

Aren't you also doing it? If so, good luck to you too!

WilliamHYZhang commented:
Nice! Yeah, I'm also doing it :)

I also left some notes in my previous comment about some file additions, please take a look if you have time.

Usually 1 file for each split is enough as to reduce space, is there a specific reason the dataset has 4 files for each split?

Also, tests should pass regardless of whether or not there is a test split. An example of this is one of my PR #1276, where the DIV2K Dataset only had train and validation splits similar to yours. The test just make sure everything is formatted correctly and working properly. I think the tests aren't passing because you're using temporary files for the `.JPEG`s, I elaborated more on that in my previous comment as well.

rickwierenga commented:
Here are the Gists:

* [160px](https://gist.github.com/rickwierenga/117d372199cacd2ce3efd325214d2a6b)
* [320px](https://gist.github.com/rickwierenga/f02a8db4bb8c7dd33900432597429109)
* [full size](https://gist.github.com/rickwierenga/603a488e4f15da5648af7952165f9ad6)

I added all metafiles in https://github.com/tensorflow/datasets/pull/1326/commits/117186ada4e87c67dbccbeb83dc78f669b0e8397 and deleted the old images in https://github.com/tensorflow/datasets/pull/1326/commits/e97f12e624132a9554d65525bb1be41fe6921af3. I copied the files from https://github.com/tensorflow/datasets/tree/master/tensorflow_datasets/testing/test_data/fake_examples/imagenette because Imagenette partly overlaps with Imagewang. I guess these files can also be deleted, but that should be in another PR. I believe the sizes are set by the dataset.

I will look at the tests next.

rickwierenga commented:
Turns out I wrote the tests incorrectly... Should be fixed in https://github.com/tensorflow/datasets/pull/1326/commits/6db5ce5e5a907efdb770a8eef41a7ee492d53813.

WilliamHYZhang commented:
Great! Can't believe we both missed that silly mistake... haha

Even with the temporary files, now that the tests are fixed they should be passing. However when running `yield fname, record`, `fname` might still be the name of that temporary file and not the desired file name. I don't know whether that needs to be changed. Other than that, once you add the fake data generation file into `testing` you should be good to go.

Hopefully a mentor will review this soon and approve your task!

rickwierenga commented:
Loading the images seems to work fine. 

> Other than that, once you add the fake data generation file into `testing` you should be good to go.

I'm not sure what you mean by that.

WilliamHYZhang commented:
In order to reduce size of files and prevent any copyright issues, almost always for each dataset a python script is made to create small sized random noise images as the testing files (see the files in `tensorflow_datasets/testing`), example files are there. For example, instead of the KB sizes of your images, a simple 1x1 image generated would only be a couple of bytes.

WilliamHYZhang commented:
Just received notice that the `dataset.info` files are autogenerated, only the gists are required and the files in metadata can be removed from this PR.

> The tensorflow_datasets/testing/metadata is automatically created by us from the generated datasets at head. We want to keep it automatically generated and not rely on the user PR for multiple reason (ex: avoid incompatibilities if we update the dataset_info format but the PR was created in an old version).

rickwierenga commented:
I hope this script is OK, not sure about the structure. I tried to mimic [MNIST](https://github.com/tensorflow/datasets/blob/master/tensorflow_datasets/testing/mnist.py) to keep the style consistent.

WilliamHYZhang commented on tensorflow_datasets/testing/imagewang.py:
```diff
@@ -0,0 +1,81 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+r"""Generate testing images for Imagewang. 
```
Pretty sure the lone r can be removed here.

rickwierenga replied:
Nearly all files have this so I'll keep it for consistency.

WilliamHYZhang commented on tensorflow_datasets/testing/imagewang.py:
```diff
@@ -0,0 +1,81 @@
+                    "Path to tensorflow_datasets directory")
+FLAGS = flags.FLAGS
+
+_IMAGE_FILENAME = 'test_image'
```
I think it would be better if the images had different file names styled similarly to what they're named in the dataset.

rickwierenga replied:
Not sure about that. File names shouldn't make a difference I think.

WilliamHYZhang commented on tensorflow_datasets/testing/imagewang.py:
```diff
@@ -0,0 +1,81 @@
+
+def write_image_file(filename):
+  with tf.io.gfile.GFile(filename, "wb") as f:
+    f.write(b"1" * 16)  # header
```
Consider removing comment.

WilliamHYZhang commented on tensorflow_datasets/testing/imagewang.py:
```diff
@@ -0,0 +1,81 @@
+  for size in _SIZES:
+    output_dir = imagewang_label_dir(size)
+    test_utils.remake_dir(output_dir)
+    write_image_file(os.path.join(output_dir, _IMAGE_FILENAME))
```
Does this put only one file in the subdirectory for the corresponding size? What about the test/validation splits?

cyfra commented on tensorflow_datasets/image/imagewang.py:
```diff
@@ -0,0 +1,140 @@
+        ImagewangConfig(
+            name=size,
+            size=size,
+            description=_DESCRIPTION.format(size=size)))
```
We have descriptions in two places:

the one in DatasetInfo (line 94): should be long and about whole dataset.
The one for the config - should be shorter and basically only chat about what this configuration is about (rather than about the whole dataset again).

cyfra commented on tensorflow_datasets/image/imagewang.py:
```diff
@@ -0,0 +1,140 @@
+class Imagewang(tfds.core.GeneratorBasedBuilder):
+  """ Imagewang contains Imagenette and Imagewoof combined. """
+
+  VERSION = tfds.core.Version("2.0.0")
```
can be removed (as you have VERSION on each config anyway)

cyfra commented on tensorflow_datasets/image/imagewang.py:
```diff
@@ -0,0 +1,140 @@
+            "label": tfds.features.ClassLabel(names_file=names_file)
+        }),
+        supervised_keys=("image", "label"),
+        homepage='https://dataset-homepage://github.com/fastai/imagenette',
```
looks like an error in the URL

rickwierenga replied:
This is actually the correct URL as all Imagenette, Imagewoof and Imagewang are all in the same repository.

cyfra replied:
I meant having "://" in the middle of the url ;-)
Shouldn't it be just: "https://github.com/fastai/imagenette"

rickwierenga replied:
😬

rickwierenga commented:
Thanks for the review @cyfra!

rickwierenga commented:
I'm sorry for the URL issue. I will read the code more carefully in the future.

cyfra commented on tensorflow_datasets/image/imagewang.py:
```diff
@@ -0,0 +1,143 @@
+
+import os
+
+import tensorflow as tf
```
Replace with 
import tensorflow.compat.v2 as tf

cyfra commented:
looks good - but test fake files are missing. Please generate & add them and then I can merge the PR.

rickwierenga commented:
Thanks @cyfra.

I generated the files with:

```sh
python tensorflow_datasets/testing/imagewang.py --tfds_dir tensorflow_datasets
```

WilliamHYZhang commented:
@rickwierenga could you resolve the merge conflict?

rickwierenga commented:
I think it should be fixed. I copied the current release notes from this repo and added the Imagewang line to it.

cyfra commented:
@rickwierenga - I don't see any JPEG files generated - is that expected?

rickwierenga commented:
@cyfra See https://github.com/tensorflow/datasets/pull/1326/commits/deac2675945e14bafd426b50f4e332d6213588f4.

cyfra commented:
Are the tests passing ? In the code you have:
for fpath in tf.io.gfile.glob(os.path.join(datapath, label, "*.JPEG")):

And there are no JPEG files in your fake files.

rickwierenga commented:
Test pass now. I'm sorry I did not test that. Let me know if you need anything else.

cyfra commented:
@rickwierenga - the tests are still failing with:

KeyError: u'label'

your test files are in directory called 'label' - that your code later uses as 'label' feature.
And this name must be in the dictionary - taken from names_file (as it is type tfds.features.ClassLabel) 

So the TL;DR: your fake_examples should have files like imagewang-160/train/n03394916/test_image.JPEG etc.

rickwierenga commented:
@cyfra, I forgot to `pip install -e .`. The tests should pass now.
